### PR TITLE
Raise error if loss can't be calculated - ViT MIM 

### LINKED ISSUE #23832: In ViTForMaskedImageModeling, you will receive a reconstructed_pixel_values which shape is different with input when model.config.patch_size is not 16. This further triggers an error about loss when patch_size is not 16 and bool_masked_pos is not None.
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-3.10.0-1160.90.1.el7.x86_64-x86_64-with-glibc2.35
- Python version: 3.9.16
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 1.13.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Use ViTForMaskedImageModeling;
Set model.config.patch_size except 16;
Get error.

### Expected behavior

In ViTForMaskedImageModeling, you will receive a reconstructed_pixel_values which shape is different with input when model.config.patch_size is not 16. This further triggers an error about loss when patch_size is not 16 and bool_masked_pos is not None.

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Currently, `ViTForMaskedImageModeling` will fail when calculating the reconstruction loss if a patch size other than 16 is chosen. This is because the decoder head is parametrized by `config.encoder_stride`, which controls the resolution of the upsampled image. 

By default, `config.patch_size = config.encoder_stride = 16`. If a user updates the patch size but not the encoder stride to match, the reconstructed image will have a different resolution. 

This PR adds a warning for the user before the forward pass, explaining why the loss calculation won't work.

Fixes #23832

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks!
